src/systems/physics/Physics.cc

```diff
@@ -1923,34 +1978,47 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
             *(mutableModelPose) = components::Pose(
-                math::eigen3::convert(worldPose) *
+                worldPoseMath3d *
                 linkPoseFromTopLevelModel.Inverse());
 
             _ecm.SetChanged(topLevelModelEnt, components::Pose::typeId,
                 ComponentState::PeriodicChange);
-          }
-          else
-          {
-            // Compute the relative pose of this link from the top level model
-            // first get the world pose of the top level model
-            auto worldComp =
-                _ecm.Component<components::ParentEntity>(topLevelModelEnt);
-            // if the worldComp is a nullptr, something is wrong with ECS
-            if (!worldComp)
+
+            // mark this model pose as changed and update all links in this
+            // model that have a pose stored w.r.t. the model pose
+            modelPoseChanged.insert(topLevelModelEnt);
+            auto modelLinks = gazebo::allModelLinks(topLevelModelEnt, _ecm);
+            for (const auto &modelLink : modelLinks)
             {
-              ignerr << "The parent component of " << topLevelModelEnt
-                     << " could not be found. This should never happen!\n";
-              return true;
+              // skip the canonical link since it was just updated
+              if (modelLink == _entity)
+                continue;
+
+              auto modelLinkPhys = this->entityLinkMap.Get(modelLink);
+              if (nullptr == modelLinkPhys)
+              {
+                ignerr << "Internal error: link [" << _entity
+                       << "] not in entity map" << std::endl;
+                continue;
+              }
+              auto modelLinkFrameData =
+                modelLinkPhys->FrameDataRelativeToWorld();
+              this->UpdateRelativeLinkPose(topLevelModelEnt, modelLink,
+                  _ecm.Component<components::ParentEntity>(modelLink),
+                  _ecm.Component<components::Pose>(modelLink),
+                  ignition::math::eigen3::convert(modelLinkFrameData.pose),
+                  _ecm);
+
+              // mark this link as changed due to a model pose update
+              linkPoseChanged.insert(modelLink);
             }
-            math::Pose3d parentWorldPose =
-                this->RelativePose(worldComp->Data(), _parent->Data(), _ecm);
-
-            // Unlike canonical links, pose of regular links can move relative
-            // to the parent. Same for links inside nested models.
-            *_pose = components::Pose(math::eigen3::convert(worldPose) +
-                                      parentWorldPose.Inverse());
-            _ecm.SetChanged(_entity, components::Pose::typeId,
-                ComponentState::PeriodicChange);
+          }
+          // don't update the link's pose if it was already updated due to a
+          // model pose upate
+          else if (linkPoseChanged.find(_entity) == linkPoseChanged.end())
```

Comment:
If the canonical link of the model is not the first link in the model (which can happen if the `canonical_link` attribute is used), the non-canonical links get updated first and then when it's time for the canonical link to get updated, all the non-canonical links get updated again. This fine in terms of correctness, but it might be a performance hit. If we want to fix that, we should update just the canonical links first and then the rest of the links. 

In fact, if we are looping through all the links of a model above, I don't know if there's even a need for this `UpdateRelativeLinkPose` here.

Reply:
I went ahead and updated the canonical links first in https://github.com/ignitionrobotics/ign-gazebo/pull/876/commits/862f0b7dc93b332b8890621e2b4d982fe51688ce. It's not the prettiest solution since we don't have the `ModelCanonicalLink` component that was introduced in #685, but this should be easier to do once we forward-port to Edifice since this component is available starting in Edifice.

As a result of this change, the `UpdateRelativeLinkPose` method is no longer needed, so I removed it.